src/smt/theory_str_noodler/formula.cpp

```diff
@@ -1,7 +1,160 @@
+            case LenFormulaType::FALSE:
+                return;
+            case LenFormulaType::LEAF: {
+                if (root.atom_val.is_variable()) {
+                    const BasicTerm& var = root.atom_val;
+                    if (!quantified_vars.contains(var)) {
+                        free_vars.insert(var);
+                    }
+                }
+                return;
+            }
+            case LenFormulaType::NOT:
+                collect_free_vars_rec(root.succ.at(0), free_vars, quantified_vars);
+                return;
+            case LenFormulaType::LEQ:
+            case LenFormulaType::LT:
+            case LenFormulaType::EQ:
+            case LenFormulaType::NEQ:
+                collect_free_vars_rec(root.succ.at(0), free_vars, quantified_vars);
+                collect_free_vars_rec(root.succ.at(1), free_vars, quantified_vars);
+                return;
+            case LenFormulaType::PLUS:
+            case LenFormulaType::MINUS:
+            case LenFormulaType::TIMES:
+            case LenFormulaType::AND:
+            case LenFormulaType::OR: {
+                for (const auto& child : root.succ) {
+                    collect_free_vars_rec(child, free_vars, quantified_vars);
+                }
+                return;
+            }
+            case LenFormulaType::FORALL:
+            case LenFormulaType::EXISTS: {
+                auto children_iterator = root.succ.begin();
+                const LenNode& quantified_var_node = *children_iterator;
+                const BasicTerm quantified_var = quantified_var_node.atom_val;
+                children_iterator++;
+
+                auto [target_bucket, did_emplace_happen] = quantified_vars.emplace(quantified_var);
+                assert (did_emplace_happen);  // We want all quantified vars to be unique, at least for now
+
+                for (; children_iterator != root.succ.end(); children_iterator++) {
+                    collect_free_vars_rec(*children_iterator, free_vars, quantified_vars);
+                }
+
+                quantified_vars.erase(quantified_var);
+
+                return;
+            }
+            default:
+                UNREACHABLE();
+        }
+    }
+
+    LenNodePrecision get_resulting_precision_for_conjunction(const LenNodePrecision& p0, const LenNodePrecision& p1) {
+        // If one of the formulae is an underapproximation, then when we find that the resulting conjunction has no solutions,
+        // then we cannot say that for sure - we might have lost some solutions due to imprecision of one of the conjuncts.
+        if (p0 == LenNodePrecision::UNDERAPPROX || p1 == LenNodePrecision::UNDERAPPROX) return LenNodePrecision::UNDERAPPROX;
+
+        // If p0 is precise, and p1 is an overapproximation, then taking a conjunction will preserve all the precise solutions
+        if (p0 == LenNodePrecision::PRECISE || p1 == LenNodePrecision::PRECISE) return LenNodePrecision::PRECISE;
```

Comment:
Is this correct? Let's say the precise formula is `true`, so the conjunction returns the other formula which is overapproximation, so it should still be overapproximation, no?

Reply:
I think so. I have fixed it.